src/server/dflycmd.cc

```diff
@@ -59,6 +64,92 @@ struct TransactionGuard {
   Transaction* t;
 };
 
+// Buffered single-shard journal streamer that listens for journal changes with a
+// journal listener and writes them to a destination sink in a separate fiber.
+class JournalStreamer {
+ public:
+  JournalStreamer(io::Sink* dest, journal::Journal* journal, Context* cntx)
+      : cntx_{cntx}, journal_{journal}, mu_{}, cv_{}, buffer_{}, dest_{dest}, write_fb_{},
+        writer_{&buffer_}, journal_cb_id_{0} {
+  }
+
+  // Register journal listener and start writer in fiber.
+  void Start();
+
+  // Must be called on context cancellation for unblocking
+  // and manual cleanup.
+  void Cancel();
+
+ private:
+  // Writer fiber that steals buffer contents and writes them to dest.
+  void WriterFb();
+
+ private:
+  Context* cntx_;
+  journal::Journal* journal_;
+
+  ::boost::fibers::mutex mu_;               // guard buffer and condvar wakeups
+  ::boost::fibers::condition_variable cv_;  // notify new data is available or cancelled
+
+  io::StringFile buffer_;
+  io::Sink* dest_;
+
+  Fiber write_fb_;
+  JournalWriter writer_;
+
+  uint32_t journal_cb_id_;
+};
+
+void JournalStreamer::Start() {
+  write_fb_ = Fiber(&JournalStreamer::WriterFb, this);
+
+  journal_cb_id_ = journal_->RegisterOnChange([this](const journal::Entry& entry) {
+    error_code ec;
+    {
+      lock_guard lk(mu_);
+      ec = writer_.Write(entry);
+    }
+
+    if (ec)
+      cntx_->Error(ec);
+
+    cv_.notify_all();
+  });
+}
+
+void JournalStreamer::Cancel() {
+  DCHECK(cntx_->IsCancelled());  // The writer exits only in this case
+
+  journal_->UnregisterOnChange(journal_cb_id_);
+  cv_.notify_all();
+
+  if (write_fb_.IsJoinable())
+    write_fb_.Join();
+}
+
+void JournalStreamer::WriterFb() {
+  // Stash and buffer are swapped in turns.
+  std::string stash;
+
+  while (true) {
+    {
+      std::unique_lock lk(mu_);
+      cv_.wait(lk);
+      std::swap(buffer_.val, stash);
+    }
```

Comment:
Still not sure about this approach. Two issues:

1. We should have backpressure if we accumulate many large values

2. Should the writer be just a freely floating fiber? What about scheduling fairness? Maybe we should enqueue it on the shard set (progress guarantee)?

Reply:
The first issue is also directly connected with the first...

If we block on backpressure with some external help (i.e. channel filled, condvar) then what are the guarantees that the other side wakes up in time, for example if we use fair scheduling on a shard set?

In this case it would be easier to write directly from the callback fiber, so we don't have to think about scheduling. If we write directly, then it would be best to do it in small portions to lower throughput, which means that we need to use a channel + pooled blobs?

Reply:
Ideally, we want it to run often (but not too often and certainly not too rarely) and evenly, so that if we have many small consecutive transactions, we would be able to write them in small batches. This certainly requires smth more advanced in the future that would analyze the shard-set for pending executions.